debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java

```diff
@@ -0,0 +1,149 @@
+import io.debezium.schema.DataCollectionSchema;
+import io.debezium.schema.DatabaseSchema;
+import io.debezium.schema.SchemaChangeEvent;
+import io.debezium.schema.TopicSelector;
+
+/**
+ * Central dispatcher for data change and schema change events. The former will be routed to the change event queue, the
+ * latter will be routed to the {@link DatabaseSchema}. But based on the applying whitelist/blacklist configuration,
+ * events may be not be dispatched at all.
+ * <p>
+ * This router is also in charge of emitting heartbeat messages, exposing of metrics via JMX etc.
+ *
+ * @author Gunnar Morling
+ */
+public class EventDispatcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventDispatcher.class);
+
+    private final TopicSelector topicSelector;
+    private final DatabaseSchema schema;
+    private final ChangeEventQueue<Object> queue;
+
+    public EventDispatcher(TopicSelector topicSelector, DatabaseSchema schema,
+            ChangeEventQueue<Object> queue) {
+        this.topicSelector = topicSelector;
+        this.schema = schema;
+        this.queue = queue;
+    }
+
+    /**
+     * Dispatches one or more {@link DataChangeEvent}s. If the given data collection is included in the currently
+     * captured set of collections, the given emitter will be invoked, so it can emit one or more events (in the common
+     * case, one event will be emitted, but e.g. in case of PK updates, it may be a deletion and a creation event). The
+     * receiving coordinator creates {@link SourceRecord}s for all emitted events and passes them to the given
+     * {@link ChangeEventCreator} for converting them into data change events.
+     */
+    public void dispatchDataChangeEvent(DataCollectionId dataCollectionId, Supplier<ChangeRecordEmitter> changeRecordEmitter, ChangeEventCreator changeEventCreator) throws InterruptedException {
+        // TODO Handle Heartbeat
+
+        // TODO Handle JMX
+
+        // TODO handle filtering based on table id
+        boolean tableIncluded = true;
+
+        if(!tableIncluded) {
+            return;
+        }
+
+        DataCollectionSchema dataCollectionSchema = schema.getDataCollectionSchema(dataCollectionId);
+
+        // TODO handle as per inconsistent schema info option
+        if(dataCollectionSchema == null) {
+            throw new IllegalArgumentException("No metadata registered for captured table " + dataCollectionId);
+        }
+
+        changeRecordEmitter.get().emitChangeRecords(
+            dataCollectionSchema,
+            new ChangeRecordReceiver(dataCollectionId, changeEventCreator, dataCollectionSchema)
+        );
+    }
+
+    public void dispatchSchemaChangeEvent(TableId tableId, Supplier<SchemaChangeEventEmitter> schemaChangeEventEmitter) throws InterruptedException {
+        // TODO
+        boolean tableIncluded = true;
+
+        if(tableIncluded) {
+            schemaChangeEventEmitter.get().emitSchemaChangeEvent(new SchemaChangeEventReceiver());
```

Comment:
Are the receivers really stateful? Do you need a new instance per change?

Reply:
Good question, probably not.